sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamSortRel.java

```diff
@@ -210,33 +211,57 @@ public int getCount() {
   }
 
   private class LimitTransform<T> extends PTransform<PCollection<T>, PCollection<T>> {
+    private int startIndex;
+
+    public LimitTransform(int startIndex) {
+      this.startIndex = startIndex;
+    }
+
     @Override
     public PCollection<T> expand(PCollection<T> input) {
       Coder<T> coder = input.getCoder();
       PCollection<KV<String, T>> keyedRow =
           input.apply(WithKeys.of("DummyKey")).setCoder(KvCoder.of(StringUtf8Coder.of(), coder));
 
-      return keyedRow.apply(ParDo.of(new LimitFn<T>(getCount())));
+      return keyedRow.apply(ParDo.of(new LimitFn<T>(getCount(), startIndex)));
     }
   }
 
   private static class LimitFn<T> extends DoFn<KV<String, T>, T> {
     private final Integer limitCount;
+    private final Integer startIndex;
 
-    public LimitFn(int c) {
+    public LimitFn(int c, int s) {
       limitCount = c;
+      startIndex = s;
     }
 
     @StateId("counter")
     private final StateSpec<ValueState<Integer>> counterState = StateSpecs.value(VarIntCoder.of());
 
+    @StateId("skipped_rows")
+    private final StateSpec<ValueState<Integer>> skippedRowsState =
+        StateSpecs.value(VarIntCoder.of());
+
     @ProcessElement
     public void processElement(
-        ProcessContext context, @StateId("counter") ValueState<Integer> counterState) {
+        ProcessContext context,
+        @StateId("counter") ValueState<Integer> counterState,
+        @StateId("skipped_rows") ValueState<Integer> skippedRowsState) {
+      Integer toSkipRows;
```

Comment:
I summarized suggestions from Kenn and me. 

It is better to split this method into two parts:
1. skip row if needed
2. do standard processing.
This will make reading code much easier.

toSkipRows initialization might need some update.
```suggestion
      Integer toSkipRows = firstNonNull(skippedRowsState.read(), startIndex);
      if (toSkipRows > 0) {
          skippedRowsState.write(toSkipRows - 1);
          return;
      }
      
      int current = (counterState.read() != null ? counterState.read() : 0);
      if (current < limitCount) {
        counterState.write(current + 1);
        context.output(context.element().getValue());
      }
```